### PR TITLE
raribleIngestor

### PR DESCRIPTION
**Pull request by s29papi**
Ingestor for Rarible

### CODE REVIEW

**Comment by s29papi** [comment]
[https://github.com/floornfts/mobile-minting/issues/10](url)

**Comment by s29papi** [comment]
closes #10

**Comment by s29papi** [comment]
@chrismaddern Could you check this out

**Comment by chrismaddern** [comment]
Thank you for this submission!! 🥳 🙏 

We've gotten a little behind on reviewing these, but will be spending some time over the next 48 hours, so look out for some feedback soon! Thank you! 🙏

File: src/ingestors/rarible/index.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        mintBuilder.setCreator({
+                name: 'not available',
+                walletAddress: ownerAddress,
+        });
+
+        let totalPriceWei;
+        if (unitMintTokenAddress == '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE' || unitMintTokenAddress == '0x4200000000000000000000000000000000000006') {
+            totalPriceWei = unitMintPrice;
+        } else {
+            totalPriceWei = await getRaribleMintPriceInEth(unitMintPrice, unitMintTokenAddress);
```
Comment (chrismaddern) [inline src/ingestors/rarible/index.ts:90]: Is this common?

FWIW we don't currently support mints for mobile minting in non-native currencies (signer wallets have ETH on Ethereum network chains, and SOL on Solana)

File: src/ingestors/rarible/offchain-metadata.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+export const urlForValidRaribleCollection = async (
+    chainId: number,
+    contractAddress: string,
+): Promise<string | undefined> => {
+    const apiKey = process.env.RARIBLE_API_KEY;
+    if (!apiKey) {
+        throw new Error('RARIBLE_API_KEY is not defined in the environment variables');
```
Comment (chrismaddern) [inline src/ingestors/rarible/offchain-metadata.ts:7]: Deployment note -- requires `RARIBLE_API_KEY` to be set.

Confirming availability of this key.

File: src/ingestors/rarible/onchain-metadata.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        stopDate: new Date(parseInt(claimCondition2[0][0]) * 1000),
+    };
+};
+
+export const getRaribleMintPriceInEth = async (
+    unitMintPrice: number,
+    unitMintTokenAddress: string,
+  ): Promise<any> => {
+    const apiKey = process.env.OX_API_KEY;
+    if (!apiKey) {
```
Comment (chrismaddern) [inline src/ingestors/rarible/onchain-metadata.ts:61]: Note that we likely won't include a 0x key in the project

Is it guaranteed that the result of a conversion on 0x will yield an amount that will be accepted by the contract for the mint here?

This is cool and interesting, but as I mentioned above, likely wouldn't be supported by mobile minting today anyway.

File: src/ingestors/rarible/index.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+       mintBuilder.setMintInstructions({
+            chainId,
+            contractAddress,
+            contractMethod: 'claim',
```
Comment (chrismaddern) [inline src/ingestors/rarible/index.ts:96]: I don't see this method in the ABI?

File: src/ingestors/rarible/index.ts (hunk 1 above)
Comment (s29papi) [reply src/ingestors/rarible/index.ts:90]: We could exclude non ether mints. I observed at the time of making this that there were a handful of non ether mints mostly in degen.

**Comment by chrismaddern** [comment]
Thanks for taking the time to share this!

I tried it on a new BASE mint using:

```
yarn run dry-run rarible url https://rarible.com/collection/base/0x968ca01b5C32234F4d6Bfd44fF079BE14789bA10/drops
```

I got an RPC error output:
CleanShot 2024-07-16 at 15 50 31@2x

Any idea if I'm doing something wrong if there's an edge case in the ingestor?

(I added a Rarible API key to the environment on the test machine)

**Comment by s29papi** [comment]
two different mints, two different contract methods:

mint you are testing: https://basescan.org/address/0x968ca01b5C32234F4d6Bfd44fF079BE14789bA10#readContract

mint type my test assumptions were based of: https://basescan.org/address/0xcc3725d7e946d512a6C439E716C7C5A41cCce182#readContract

The error originates from the contract method 'sharedMetadata' which is missing in the new mint type.

The way around this would be to get the metadata stored on ipfs, and to see if it works for both. It should.

File: src/ingestors/rarible/onchain-metadata.ts (hunk 3 above)
Comment (s29papi) [reply src/ingestors/rarible/onchain-metadata.ts:61]: Sticking to only ether based mints, removes the need for an OX api key as this is used for conversion.

**Comment by chrismaddern** [comment]
Amazing, that sounds smart.

FYI — I've updated the main branch on floor/mobile-minting to add simulation to the test suite for Base.

If you:
* Merge floor/mobile-minting main into your branch
* Add `SIMULATE_DURING_TESTS=true` to your .env
* Use the `basicIngestorTests` macro to your test suite ([example here](https://github.com/floornfts/mobile-minting/blob/main/test/ingestors/fxhash.test.ts#L12))

Then... running it should show you which created mints will actually execute onchain 🥳 

One thing I noticed (beyond the fact it will only support Base, not Degen) is that the encoded parameters in the contractParams have been truncated to exponent form - e.g. 1e21, vs. a bigint string form "10000000000...". Unless the contract is explicitly expecting that, I expect that will fail.

Let me know if you'd like to land this, we're holding the bounty spot for this one if you do.

**Comment by s29papi** [comment]
Hey @chrismaddern, Could the data type for the successUrls be changed from string[] ---> [{url: "wyz.com", block: "0x452395"}, ...]. or block be added to type `MintContractOptions` or it could be the key while still retaining passing various block.

Currently the simulations blocks is a mapping of chainId to BlockHeight this is not ideal for me as each contract or url has a period of time where minting is valid and after with simulating a transaction fails, and the issue in this case being the key (chainId).

Any change that includes adding the block to the successUrls or key for blocks, it would be feasible to pass various blocks while iterating:

`
  const result = await simulateEVMTransactionWithAlchemy(mintInstructions, simulationBlocks[mintInstructions.chainId]);
`

**Comment by chrismaddern** [comment]
Hi @s29papi if you're confident that this is an overlapping blockhash issue vs. a code generation issue then I can take a look at updating the format.

I was _hoping_ that for now we'd be able to find a couple test URLs that were available in the same time window to begin, but it does seem likely we'd need to iterate on this eventually anyway, so.....

**Comment by s29papi** [comment]
If the direction for now is just mints within the same mint period, it's not a problem I would adjust to that.

**Comment by s29papi** [comment]
This should be up for review. Regarding the initial issue I raised, I provided an array of objects that include both block height and URL, along with an index that can be manually changed to test all mints with their respective block heights.

File: src/ingestors/rarible/index.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            conditionProof, 
+            ownerAddress,
+            startDate,
+            stopDate
+        } = await raribleContractMetadataGetter(chainId, contractAddress, resources.alchemy)
+        
+        mintBuilder.setName(name).setDescription(description).setFeaturedImageUrl(imageURI);
+        mintBuilder.setMintOutputContract({ chainId: chainId, address: contractAddress });
+        mintBuilder.setCreator({
+                name: 'not available',
```
Comment (chrismaddern) [inline src/ingestors/rarible/index.ts:78]: This will show up in the UI which we don't want.

It does look like rarible.org don't include this in the public API with key, but rarible.com do provide it in their API which appears accessible

```
curl 'https://rarible.com/marketplace/api/v4/profiles/list' \
  -H 'accept: */*' \
  -H 'content-type: application/json' \
  --data-raw '["0xbab964a06fbd884b0ad0efccb86f5931248858cc"]'
```

This has the name and other details which would make this much better! 🙂 

CleanShot 2024-07-23 at 09 57 21@2x

**Comment by chrismaddern** [comment]
> This should be up for review. Regarding the initial issue I raised, I provided an array of objects that include both block height and URL, along with an index that can be manually changed to test all mints with their respective block heights.

If there aren't matching mints that we can use in the same time window, I'll go update the tests and commit to main an update to allow specifying a <URL/Contract, Blockhash> pair

File: src/ingestors/rarible/index.ts (hunk 5 above)
Comment (s29papi) [reply src/ingestors/rarible/index.ts]: Hey, thanks. Been in and out of their telegram chat and couldn't get this endpoint.

**Comment by s29papi** [comment]
> > This should be up for review. Regarding the initial issue I raised, I provided an array of objects that include both block height and URL, along with an index that can be manually changed to test all mints with their respective block heights.
> 
> If there aren't matching mints that we can use in the same time window, I'll go update the tests and commit to main an update to allow specifying a <URL/Contract, Blockhash> pair

Alright, I would update my test file after you've made the changes.

**Comment by s29papi** [comment]
@chrismaddern, Took timeout to dig more into this, oops! my apologies, I was able to Find two mints which share same range for active claims within a particular block height. Would two be enough ? Although I had tested 6 mints individually which all passed at with their respective block height.

**Comment by s29papi** [comment]
Hey @chrismaddern I would like to know what's left to land this I have multiple tests passed, but this is still marked on final approach

**Comment by chrismaddern** [comment]
Hi @s29papi sorry for the delay!

I was hoping to be able to give you a quick thumbsup, but it seems one of the tests is now not passing?

CleanShot 2024-08-27 at 06 51 06@2x

Can you get that passing and then we can get it merge.

A quick spotcheck with Rarible mint looks great!

```
yarn dry-run rarible url https://rarible.com/collection/base/0x7a2e753a20ad60a931643ce87aeadb8efbf1caa1/drops
```

If you can get the tests passing again today, we can get it merged today!

**Comment by s29papi** [comment]
Hi @chrismaddern. This is fixed, and the tests pass. 

Sorry for the oversight.

timestampClaimcondition

Fixed it.

**Review by chrismaddern — APPROVED** [review]
Tests passing locally. Added RARIBLE_API_KEY to workflow and environment secrets

**Comment by chrismaddern** [comment]
@s29papi merging and sending you an email for bounty claim!! 🥳
